engine/kv_engine_cleaner.cpp

```diff
@@ -617,22 +605,177 @@ void Cleaner::AdjustThread(size_t advice_thread_num) {
+                           outdated_collections_.skiplists.size());
+  {
+    std::unique_lock<std::mutex> skiplist_lock(kv_engine_->skiplists_mu_);
+    auto iter = kv_engine_->expirable_skiplists_.begin();
+    while (!kv_engine_->expirable_skiplists_.empty() && (*iter)->HasExpired() &&
+           limited_fetch_num < max_thread_num_) {
+      auto outdated_skiplist = *iter;
+      iter = kv_engine_->expirable_skiplists_.erase(iter);
+      outdated_collections_.skiplists.emplace(outdated_skiplist);
+      limited_fetch_num++;
+    }
+  }
+
+  {
+    std::unique_lock<std::mutex> list_lock(kv_engine_->lists_mu_);
+    auto iter = kv_engine_->lists_.begin();
+    while (!kv_engine_->lists_.empty() && (*iter)->HasExpired() &&
+           limited_fetch_num < max_thread_num_) {
+      auto expired_list = *iter;
+      iter = kv_engine_->lists_.erase(iter);
+      outdated_collections_.lists.emplace(expired_list);
+      limited_fetch_num++;
+    }
+  }
+
+  {
+    std::unique_lock<std::mutex> hlist_lock(kv_engine_->hlists_mu_);
+    auto iter = kv_engine_->hash_lists_.begin();
+    while (!kv_engine_->hash_lists_.empty() && (*iter)->HasExpired() &&
+           limited_fetch_num < max_thread_num_) {
+      auto expired_hash_list = *iter;
+      iter = kv_engine_->hash_lists_.erase(iter);
+      outdated_collections_.hashlists.emplace(expired_hash_list);
+      limited_fetch_num++;
+    }
+  }
+
+  int64_t after_queue_size =
+      static_cast<int64_t>(outdated_collections_.hashlists.size() +
+                           outdated_collections_.lists.size() +
+                           outdated_collections_.skiplists.size());
+
+  double diff_size_ratio =
+      (after_queue_size - before_queue_size) / max_thread_num_;
+  if (diff_size_ratio > 0) {
+    outdated_collections_.increase_ratio =
+        outdated_collections_.increase_ratio == 0
+            ? diff_size_ratio
+            : diff_size_ratio / outdated_collections_.increase_ratio;
+  }
+  return outdated_collections_.increase_ratio;
+}
+
+void Cleaner::FetchOutdatedCollections(
+    PendingCleanRecords& pending_clean_records) {
+  Collection* outdated_collection = nullptr;
+  RecordType record_type;
+  auto min_snapshot_ts =
+      kv_engine_->version_controller_.GlobalOldestSnapshotTs();
+  {
+    std::unique_lock<SpinMutex> queue_lock(outdated_collections_.queue_mtx_);
+    if (!outdated_collection && !outdated_collections_.skiplists.empty()) {
+      auto skiplist_iter = outdated_collections_.skiplists.begin();
+      if ((*skiplist_iter)->GetTimeStamp() < min_snapshot_ts) {
+        outdated_collection = *skiplist_iter;
+        record_type = RecordType::SortedHeader;
+        outdated_collections_.skiplists.erase(skiplist_iter);
+      }
+    }
+
+    if (!outdated_collection && !outdated_collections_.lists.empty()) {
+      auto list_iter = outdated_collections_.lists.begin();
+      if ((*list_iter)->GetTimeStamp() < min_snapshot_ts) {
+        outdated_collection = *list_iter;
```

Comment:
This may overwrite the skiplist from previous branch.
Consider using 3 pointers to child class instead of 1 pointer to parent.

Reply:
will not,  it has `!outdated_collection` check. see 684 line.